### PR TITLE
Amnezia PoC

### PR DESCRIPTION
**Pull request by jmwample**
PR allowing for drop in replacement of `wireguard-go` with `amneziawg`.

In total this is pretty straight-forward - the amnezia lib uses the same external interface but has extra (optional) arguments that are added as part of the String configuration passed in on initialization. So we swap out the library paths and then add in the options and things just work.

For now, this PR is a client change only. And as none of the servers support the more advanced features of amnezia, we cannot use handshake (init or response) message padding or header remapping yet.

```
All parameters should be the same between Client and Server, except Jc - it can vary.

- Jc — (1 ≤ Jc ≤ 128); recommended range is from 3 to 10 inclusive
- Jmin — (Jmin < Jmax); recommended value is 50
- Jmax — ( Jmin < Jmax ≤ 1280); recommended value is 1000
- S1 — (S1 < 1280; S1 + 56 ≠ S2); recommended range is from 15 to 150 inclusive
- S2 — (S2 < 1280); recommended range is from 15 to 150 inclusive
- H1/H2/H3/H4 — must be unique among each other;
```

For the Parameters that we can set there is no requirement that they match or even stay constant. So we can hard code them in for now, or (if we want) randomize them on the client on connection startup using amnezia.

What do these parameters do? 
- `jc` - enables junk packets preceding the handshake init message
- `jmin` - sets the lower bound for the size of junk packets
- `jmax` - sets the upper bound for the size of junk packets

## How does Amnezia actually differ from wireguard?
1. Junk packets can be send PRECEDING the Handshake init - the number and length of these packets is configurable.
2. Junk bytes can be PREPENDED  to the handshake init message.
3. Junk bytes can be PREPENDED  to the handshake response message.
4. The four Header type values can be remapped to alternate values.

Changes 2, 3, and 4 REQUIRE both client and server to run AmneziaWG with matching configuration options S1, S2,  H1, H2, H3, and H4.

## General Question(s)

How do we want to make this available?
- library option that can be enabled in settings in the app
- separate beta build
- auto-detect locale and try amnezia
- other?

- - -
This change is [Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415)

### CODE REVIEW

File: wireguard/build-wireguard-go.sh [hunk 1]
```diff
@@ -2,39 +2,57 @@
...
+if [ $? != 0 ] ; then echo "Terminating..." >&2 ; exit 1 ; fi
+
+# Note the quotes around '$TEMP': they are essential!
+eval set -- "$TEMP"
+
+ANDROID_BUILD=false
+IOS_BUILD=false
+DOCKER_BUILD=true
+AMNEZIA_BUILD=false
+while true; do
```
Comment (pronebird) [inline wireguard/build-wireguard-go.sh:17]: Been meaning to do that. We should extract arts parsing into main :)

Comment (jmwample) [reply wireguard/build-wireguard-go.sh:17]: I can split this out into it's own little PR

**Comment by jmwample** [comment]
I am not sure why the ios build is not finding `libwg` in `/build/lib/aarch64-apple-ios`. It could be that I missed a path replacement or implied lib name in `wireguard/build-wureguard-go.sh` somewhere.

File: wireguard/build-wireguard-go.sh (hunk 1 above)
Comment (pronebird) [reply wireguard/build-wireguard-go.sh:17]: That would be great!

**Comment by pronebird** [comment]
> I am not sure why the ios build is not finding `libwg` in `/build/lib/aarch64-apple-ios`. It could be that I missed a path replacement or implied lib name in `wireguard/build-wureguard-go.sh` somewhere.

Try `ls -la /build/lib/` and see what's in there

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415)* status: 0 of 26 files reviewed, 1 unresolved discussion (waiting on @jmwample)

___
*[`nym-vpn-core/crates/nym-vpn-lib/build.rs` line 16 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415#-OAMQ3kxD19BvY_mitrn:-OAMQ3kxD19BvY_mitro:bidjudg) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/d9acac37c361d24255a65bf4e0f7e42995323133/nym-vpn-core/crates/nym-vpn-lib/build.rs#L16)):*
> ```XML
>         .expect("failed to extract build metadata");
> 
>     let manifest_path = env::var_os("CARGO_MANIFEST_DIR").expect("manifest dir is not set");
> ```

FYI Maybe you need to rebase on `main` because `crates/nym-wg-go` already provides the linker flags, so there is no need to modify `build.rs` for individual targets.

File: wireguard/build-wireguard-go.sh (hunk 1 above)
Comment (jmwample) [reply wireguard/build-wireguard-go.sh]: PR created here - https://github.com/nymtech/nym-vpn-client/pull/1432

I think it now works for the enabled build platforms

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415)* status: 0 of 41 files reviewed, all discussions resolved (waiting on @aniampio)

___
*[`wireguard/build-wireguard-go.sh` line 17 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415#gh-1819803297:-OAWtvKwAi4WtPTc-lkw:b-xe6ih) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/d9acac37c361d24255a65bf4e0f7e42995323133/wireguard/build-wireguard-go.sh#L17)):*
<details><summary><i>Previously, jmwample (Jack Wampler) wrote…</i></summary><blockquote>

PR created here - https://github.com/nymtech/nym-vpn-client/pull/1432

I think it now works for the enabled build platforms
</blockquote></details>

Thank you for that!

**Review by pronebird — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415)* status: 0 of 41 files reviewed, 1 unresolved discussion (waiting on @aniampio and @jmwample)

*[a discussion](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415#-OAWxA559dmr2ujqiWz_:-OAWxA559dmr2ujqiWza:b-kbym0c) (no related file):*
Do you think there is a way we can maintain the same libwg codebase without copying all code and support amnesia at the same time? 

I am not very familiar with Go toolchain. I know there things like tags for conditional compilation. 

For imports, it seems that Amnezia is merely a mirror of `golang.zx2c4.com/wireguard` repo so maybe we could rewrite the repo URL to amnezia in `go.mod` while on the outside maintain the same imports from zx2c4.com. 

I am not sure what the best way to tackle this, but I'd like to avoid copying the same code for as much as possible because there will be patches to the libwg and the last thing we want is to sync between libwg and amnezia fork, if possible of course.

___

**Comment by jmwample** [comment]
> Do you think there is a way we can maintain the same libwg codebase without copying all code and support amnesia at the same time?

This is what I tried to do first. However, I ran into trouble because the go `replace` directive for `go.mod` only allows you to replace a library with a library with the same name (`"wireguard-go" != "amneziawg-go"`). At the same time, the rust compiler doesn't have a pre-processor, so you can't use tags to swap library paths like you can in rust -- build tags are used to swap out entire files.

So we could use build tags and only copy the files with references to `golang.zx2c4.com`, or we could use a patch script and search/replace those import paths. OR we could just use the amnezia library and get rid  of the wireguard-go version since it is a fork with minimal changes that by default does the exact same thing.

**Review by pronebird — CHANGES_REQUESTED** [review]
> Reviewed 1 of 20 files at r2.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415)* status: 1 of 54 files reviewed, 3 unresolved discussions (waiting on @aniampio and @jmwample)

___
*[`nym-vpn-core/crates/nym-vpnd/build.rs` line 8 at r3](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415#-OBA5fmA0VMAL5yD-Ow2:-OBA5fmA0VMAL5yD-Ow3:bid0s2f) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/8ec3cb10322a8c6f38655bc16140432725060405/nym-vpn-core/crates/nym-vpnd/build.rs#L8)):*
> ```XML
> 
> fn main() -> Result<(), Box<dyn std::error::Error>> {
>     let manifest_path = env::var_os("CARGO_MANIFEST_DIR").expect("manifest dir is not set");
> ```

Can we please revert these changes. `nym-wg-go` already does that.

**Review by pronebird — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415)* status: 1 of 54 files reviewed, 4 unresolved discussions (waiting on @aniampio and @jmwample)

___
*[`.github/workflows/ci-nym-vpn-core-ios.yml` line 49 at r3](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415#-OBA5xTS0kMsvB5wg_Mk:-OBA5xTS0kMsvB5wg_Ml:b-56fy28) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/8ec3cb10322a8c6f38655bc16140432725060405/.github/workflows/ci-nym-vpn-core-ios.yml#L49)):*
> ```YAML
>           repo-token: ${{ secrets.GITHUB_TOKEN }}
> 
>       - name: Install script dependencies
> ```

I guess we no longer need it

**Review by pronebird — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415)* status: 1 of 54 files reviewed, 5 unresolved discussions (waiting on @aniampio and @jmwample)

___
*[`wireguard/build-wireguard-go.sh` line 262 at r3](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415#-OBA62siCqxUrv4RLjCz:-OBA62siCqxUrv4RLjD-:b-5i3kic) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/8ec3cb10322a8c6f38655bc16140432725060405/wireguard/build-wireguard-go.sh#L262)):*
> ```Shell
> }
> 
> LIB_DIR="libwg"
> ```

LIB_DIR is defined at the top. Rebasing issues?

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415)* status: 1 of 54 files reviewed, 6 unresolved discussions (waiting on @aniampio and @jmwample)

___
*[`nym-vpn-core/crates/nym-gateway-probe/README.md` line 12 at r3](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415#-OBA6OXM444nKpPRPUGS:-OBA6OXM444nKpPRPUGT:b-uy41pb) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/8ec3cb10322a8c6f38655bc16140432725060405/nym-vpn-core/crates/nym-gateway-probe/README.md#L12)):*
> ```Markdown
> preffered platform.
> 
> Install required dependencies
> ```

Nit: this should really be a part of markdown document in the root of the nym-vpn-core because it defines general requirements for the entire project. Except maybe clang?

**Review by pronebird — COMMENTED** [review]
> Reviewed 1 of 20 files at r2, 1 of 37 files at r3.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415)* status: 3 of 54 files reviewed, 5 unresolved discussions (waiting on @aniampio and @jmwample)

**Review by pronebird — COMMENTED** [review]
> Reviewed 1 of 37 files at r3.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415)* status: 4 of 54 files reviewed, 5 unresolved discussions (waiting on @aniampio and @jmwample)

**Review by pronebird — COMMENTED** [review]
> Reviewed 1 of 37 files at r3.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415)* status: 5 of 54 files reviewed, 5 unresolved discussions (waiting on @aniampio and @jmwample)

**Review by pronebird — CHANGES_REQUESTED** [review]
> Reviewed 1 of 26 files at r1, 10 of 20 files at r2, 1 of 37 files at r3.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415)* status: 17 of 54 files reviewed, 7 unresolved discussions (waiting on @aniampio and @jmwample)

___
*[`.github/workflows/ci-nym-vpn-core-macos.yml` line 49 at r3](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415#-OBAPqNu3hkgGC_ObUJE:-OBAPqNu3hkgGC_ObUJF:buo9nmt) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/8ec3cb10322a8c6f38655bc16140432725060405/.github/workflows/ci-nym-vpn-core-macos.yml#L49)):*
> ```YAML
> 
>       - name: Install script dependencies
>         run: brew install gnu-getopt
> ```

Same here
___
*[`nym-vpn-core/crates/nym-gateway-probe/netstack_ping/go.mod` line 5 at r3](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415#-OBAQ0TA6Ule-BKtxFko:-OBAQ0TA6Ule-BKtxFkp:b-nbtbi9) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/8ec3cb10322a8c6f38655bc16140432725060405/nym-vpn-core/crates/nym-gateway-probe/netstack_ping/go.mod#L5)):*
> ```XML
> go 1.22.3
> 
> toolchain go1.23.1
> ```

It looks like toolchain is set to go1.23 but go itself is still at 1.22. I am certain what that means in practice.

**Review by pronebird — COMMENTED** [review]
> Reviewed 4 of 20 files at r2, 33 of 37 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415)* status: all files reviewed, 7 unresolved discussions (waiting on @aniampio and @jmwample)

**Review by jmwample — COMMENTED** [review]
I have gone through and fixed open tasks. However, I believe current intention is to close this PR and make the amnezia PoC its own branch with releases parallel to `main` for the time being.

> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415#-:-OCKNjtT-3x6l4cvbrt-:b-huhjwc)* status: 19 of 57 files reviewed, 6 unresolved discussions (waiting on @aniampio and @pronebird)

___
*[`nym-vpn-core/crates/nym-gateway-probe/netstack_ping/go.mod` line 5 at r3](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415#-OBAQ0TA6Ule-BKtxFko:-OCKNBgo9cOBE_ENPOCZ:b-kmngy3) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/8ec3cb10322a8c6f38655bc16140432725060405/nym-vpn-core/crates/nym-gateway-probe/netstack_ping/go.mod#L5)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

It looks like toolchain is set to go1.23 but go itself is still at 1.22. I am certain what that means in practice.
</blockquote></details>

The go.mod syntax is such that this means "for this package 1.22.3 is the minimum golang version supported, but 1.23.1 is the suggested version". 
___
*[`nym-vpn-core/crates/nym-vpn-lib/build.rs` line 16 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415#-OAMQ3kxD19BvY_mitrn:-OCKMxlfBFSLVvemzFrn:b-896fix) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/d9acac37c361d24255a65bf4e0f7e42995323133/nym-vpn-core/crates/nym-vpn-lib/build.rs#L16)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

FYI Maybe you need to rebase on `main` because `crates/nym-wg-go` already provides the linker flags, so there is no need to modify `build.rs` for individual targets.
</blockquote></details>

Done.
___
*[`wireguard/build-wireguard-go.sh` line 262 at r3](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415#-OBA62siCqxUrv4RLjCz:-OCKN-GC5j6Fq0YLl0Vk:b-896fix) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/8ec3cb10322a8c6f38655bc16140432725060405/wireguard/build-wireguard-go.sh#L262)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

LIB_DIR is defined at the top. Rebasing issues?
</blockquote></details>

Done.
___
*[`.github/workflows/ci-nym-vpn-core-ios.yml` line 49 at r3](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415#-OBA5xTS0kMsvB5wg_Mk:-OCKN0ht1a87owcnPVGf:b-896fix) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/8ec3cb10322a8c6f38655bc16140432725060405/.github/workflows/ci-nym-vpn-core-ios.yml#L49)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

I guess we no longer need it
</blockquote></details>

Done.
___
*[`.github/workflows/ci-nym-vpn-core-macos.yml` line 49 at r3](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415#-OBAPqNu3hkgGC_ObUJE:-OCKN1ER3Wf5c4-ouyoL:b-896fix) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/8ec3cb10322a8c6f38655bc16140432725060405/.github/workflows/ci-nym-vpn-core-macos.yml#L49)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Same here
</blockquote></details>

Done.
___
*[`nym-vpn-core/crates/nym-vpnd/build.rs` line 8 at r3](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415#-OBA5fmA0VMAL5yD-Ow2:-OCKOtgw-zH5bamNssAw:b-896fix) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/8ec3cb10322a8c6f38655bc16140432725060405/nym-vpn-core/crates/nym-vpnd/build.rs#L8)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Can we please revert these changes. `nym-wg-go` already does that.
</blockquote></details>

Done.

**Review by pronebird — APPROVED** [review]
Yeah that's probably the way to go for now.

> Reviewed 31 of 38 files at r4.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1415#-:-OCXcd1P-rJxhAz6o7xH:by9kcoq)* status: 50 of 57 files reviewed, all discussions resolved (waiting on @aniampio)

**Pull request closed without merging**